samples/net/ptp/master/src/ptp_master.c

```diff
@@ -0,0 +1,388 @@
+#include <net/net_pkt.h>
+#include <net/ptp.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zephyr.h>
+
+#include "ptp_master.h"
+
+static const struct net_eth_addr ptp_mcast_eth_addr = { { 0x01, 0x1b, 0x19, 0x00, 0x00, 0x00 } };
+
+struct ptp_master_state {
+	struct net_if *iface;
+	const struct device *clk;
+	struct ptp_port_identity port_identity;
+	struct net_context *context;
+	struct net_pkt *sync_pkt;
+	struct net_if_timestamp_cb sync_sent_cb;
+	uint16_t announce_seq_id;
+	uint16_t sync_seq_id;
+	uint8_t domain_number;
+	uint8_t log_announce_interval;
+	uint8_t log_sync_interval;
+	uint8_t log_min_delay_req_interval;
+};
+
+static struct ptp_master_state state;
+
+static void ptp_announce_sender(void);
+static void ptp_sync_sender(void);
+
+K_THREAD_DEFINE(ptp_announce_sender_thread_id, 1024, ptp_announce_sender, NULL, NULL, NULL,
+	K_PRIO_PREEMPT(8), 0, K_TICKS_FOREVER);
+K_THREAD_DEFINE(ptp_sync_sender_thread_id, 1024, ptp_sync_sender, NULL, NULL, NULL,
+	K_PRIO_PREEMPT(8), 0, K_TICKS_FOREVER);
+
+static void handle_ptp_sync_sent(struct net_pkt *pkt);
+static void send_ptp_follow_up(struct net_ptp_time *ts);
+
+static void send_ptp_announce(void)
+{
+	/* prepare body */
+	struct ptp_announce_body ptp_body;
+
+	memset(&ptp_body, 0, sizeof(ptp_body));
+	ptp_body.grandmaster_clock_quality.clock_class = 248;
+	ptp_body.grandmaster_clock_quality.clock_accuracy = 0xfe;
+	ptp_body.grandmaster_clock_quality.offset_scaled_log_var = 0xffff;
+	memcpy(&ptp_body.grandmaster_identity, &state.port_identity.clock_identity,
+		sizeof(ptp_body.grandmaster_identity));
+	ptp_body.time_source = 0xa0;
+
+	/* prepare header */
+	struct ptp_header ptp_hdr;
+
+	memset(&ptp_hdr, 0, sizeof(ptp_hdr));
+	ptp_hdr.message_type = PTP_MESSAGE_TYPE_ANNOUNCE;
+	ptp_hdr.version_ptp = 2;
+	ptp_hdr.minor_version_ptp = 0;
+	ptp_hdr.message_length = sizeof(ptp_hdr) + sizeof(ptp_body);
+	ptp_hdr.domain_number = state.domain_number;
+	ptp_hdr.flag_field = PTP_FLAG_PTP_TIMESCALE;
+	memcpy(&ptp_hdr.source_port_identity, &state.port_identity,
+		sizeof(ptp_hdr.source_port_identity));
+	ptp_hdr.sequence_id = ++state.announce_seq_id;
+	ptp_hdr.log_message_interval = state.log_announce_interval;
+
+	/* allocate packet with PTP settings */
+	struct net_pkt *pkt = net_pkt_alloc_with_buffer(state.iface, ptp_hdr.message_length,
+		AF_UNSPEC, 0, K_NO_WAIT);
+
+	net_pkt_set_ptp(pkt, true);
+	net_pkt_set_priority(pkt, NET_PRIORITY_CA);
+
+	/* set src mac addr */
+	struct net_linkaddr *iface_addr = net_if_get_link_addr(state.iface);
+
+	net_pkt_lladdr_src(pkt)->addr = iface_addr->addr;
+	net_pkt_lladdr_src(pkt)->len = iface_addr->len;
+
+	/* set dst mac addr */
+	net_pkt_lladdr_dst(pkt)->addr = (uint8_t *)&ptp_mcast_eth_addr;
+	net_pkt_lladdr_dst(pkt)->len = sizeof(ptp_mcast_eth_addr);
+
+	/* write packet PTP data */
+	net_pkt_write(pkt, &ptp_hdr, sizeof(ptp_hdr));
+	net_pkt_write(pkt, &ptp_body, sizeof(ptp_body));
+
+	/* enqueue for sending */
+	net_if_queue_tx(state.iface, pkt);
```

Comment:
Samples shouldn't queue packets directly, if you decide to use `net_context` API in the sample you should rather stick to this on the TX path as well.

Reply:
It's not important in this line, but in the next function I need to capture transmission timestamp using `net_if_register_timestamp_cb` on a packet - can I accomplish the same thing with `net_context_send()` or another API?